Fix conflicts in src/backend/storage/lmgr/lock.c

RekGRpth opened the pull request:
Fix conflicts in src/backend/storage/lmgr/lock.c

1) Commit 15ef6ff in src/backend/storage/lmgr/lock.c added a new assert with a
comment in the LockAcquireExtended function, and commit 72e78d8 added another
assert with a comment below it. Earlier GPDB-specific commits had already added
GPDB-specific code in the same location.

2) Commit c096a80 in src/backend/storage/lmgr/lock.c changed the return type
from int to bool in the LockCheckConflicts function, an earlier commit 15ef6ff
had already done this.

3) Commit 85f6b49 added a conflict even for group members, and commit 3ba59cc
added a conflict for pages in the last condition, while the earlier commit
38d8815 had already separated the function code for the MPP session and single
process separately. To reduce future conflicts, separate the GPDB-specific code
into a separate if branch.

RekGRpth committed:
Fix conflicts in src/backend/storage/lmgr/lock.c

1) Commit 15ef6ff in src/backend/storage/lmgr/lock.c added a new assert with a
comment in the LockAcquireExtended function, and commit 72e78d8 added another
assert with a comment below it. Earlier GPDB-specific commits had already added
GPDB-specific code in the same location.

2) Commit c096a80 in src/backend/storage/lmgr/lock.c changed the return type
from int to bool in the LockCheckConflicts function, and commit 85f6b49 added a
conflict even for group members, and commit 3ba59cc added a conflict for pages
in the last condition, while earlier commit 38d8815 already added this.

RekGRpth committed:
fix

dkovalev1 commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -1684,132 +1737,30 @@ LockCheckConflicts(LockMethod lockMethodTable,
-	 }
+	/* Nope, it's a real conflict. */
+	PROCLOCK_PRINT("LockCheckConflicts: conflicting (group)", proclock);
 
```
This extra line differs from postgres. I suggest removing it so diff with PG will not have it.

RekGRpth replied:
removed

RekGRpth committed:
rm empty line

RekGRpth committed:
Addition

Commit 15ef6ff added a new assertion to the LockAcquireExtended function in
src/backend/storage/lmgr/lock.c. This assertion is invalid on the executor when
inserting into an AO table. For example, the choose_segno_internal function
explicitly acquires an ExclusiveLock lock from LockRelationForExtension and
then calls the GetAppendOnlyEntryAuxOids function, which opens the table_open
table with an AccessShareLock lock. This leads to the same LockAcquireExtended
function, but with IsRelationExtensionLockHeld already set, which is precisely
what the new assertion prevents.

RekGRpth committed:
Revert "Addition"

This reverts commit af710066e5c8722bd832c08454133bbe9e9cc254.
